request-manager/graph/graph.go

```diff
@@ -250,26 +265,37 @@ func (g *Graph) edgesMatchesRevEdges() bool {
 	return true
 }
 
-// hasCyclesHelper determines if a graph has cycles, using DFS.
-// Precondition: start node is already in seen list.
-func (g *Graph) hasCyclesHelper(seen map[string]*Node, start *Node) bool {
+// Determines if a graph has cycles, using dfs. inPath is a list of nodes
+// that are in the current traversal path. visited is a list of nodes that
+// are guaranteed to not be involved in any cycles.
+// precondition: start node is already in inPath list
+func (g *Graph) hasCyclesHelper(inPath map[string]*Node, visited map[string]struct{}, start *Node) bool {
 	for _, next := range g.GetNext(start) {
+		// If we've already determined the next node is not involved in
+		// any cycles, return false
+		if _, ok := visited[next.Id]; ok {
+			return false
```

Comment:
This needs to be just a `continue` instead of a return.

See this [example graph](https://docs.google.com/drawings/d/15R2qd9sbC1U7e_MBrE2qWj5LLYP48KkbPih-lydh_vU/edit?usp=sharing). Suppose `hasCyclesHelper` is first called on node A — it will visit node C, add that to the `visited` map since there are no cycles, and then return. Later `hasCyclesHelper` is called on node B. It loops through the nodes "after" B (i.e. C + D) — let's say C is done first. Since C has already been visited, it returns here saying "there are no cycles", missing the cycle between nodes D + E.

Reply:
Great catch. Fixed